### PR TITLE
fix(core): mark learnable softmax_offset as tensor parallel for clip norm

### PR DESCRIPTION
**Pull request by dundysm**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

Fixes Finding 8 of #7452 only: mark learnable attention-sink `softmax_offset` as tensor-model-parallel so clip-norm includes every TP rank's head-sharded sink grads.

@ezyang @wujingyue finding 8 fix ready for review; verified with the in-issue cpu/gloo repro.

## Issue tracking

Linked issue: Fixes Finding 8 of #7452 (does not claim to fix the whole issue)

## Root cause

With `softmax_type="learnable"`, `softmax_offset` is sized per local TP heads and `sharded_state_dict` already shards dim 0, but the parameter was never marked `tensor_model_parallel`. `param_is_not_tensor_parallel_duplicate` then keeps sink grads only on TP rank 0 for clip-norm, undercounting ‖g‖₂. Same symptom family as merged #5916 (expert grad undercount).

## Approach

1. `DotProductAttention`: after creating the learnable `softmax_offset`, call `tensor_parallel.set_tensor_model_parallel_attributes(self.softmax_offset, True, 0, 1)`.
2. `TEDotProductAttention`: after `super().__init__`, if TE created a learnable `softmax_offset` without TP attributes, set the same attributes (set-if-missing).
3. Unit test asserts the TP attributes on a mocked TP=2 `DotProductAttention`.

## Verification

Post-fix CPU/Gloo repro (adapted from the Finding 8 script in #7452):

```bash
PYTHONPATH=. torchrun --standalone --nproc-per-node=2 verify_08_attention_sink_grad_norm.py
```

Result (DotProductAttention path; TE not installed locally):

- observed: 5.0
- reference: 5.0
- `VERIFIED attention_sink_grad_norm_fixed`

Before the fix the same setup selected only rank-0's contribution and reported observed 3.0 vs reference 5.0.

Also ran a focused attribute check: `softmax_offset.tensor_model_parallel is True`, `partition_dim == 0`, `partition_stride == 1`.

### Pre-checks

- [x] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] If this PR adds or changes a GPU kernel (Triton, `jit_fuser`/`torch.compile`, CUDA extension, TE or external-library dispatch, or a scatter/index accumulation), I have added or updated its bit-exact determinism test and registered it in `tests/unit_tests/determinism/kernels/manifest.py`
- [x] I have added proper typing to my code
- [ ] I have added relevant documentation
- [ ] I have run the autoformatter.sh on my PR

### CODE REVIEW

**Comment by ezyang** [comment]
For Megatron maintainers: when I was fixing a similar class of issues in a separate codebase, I did not feel SOTA LLMs could one shot fixes at an acceptable level of quality. Usually the problem was that there was some structural problem that needed fixing at the same time.

**Comment by Connor-XY** [comment]
@wujingyue, could you take the maintainer review here, including the end-to-end clip-norm behavior for the local and Transformer Engine attention paths? The reported numerical check covers the local CPU/Gloo path; the TE path still needs coverage. That should help address the broader correctness concern raised above.

**Comment by wujingyue** [comment]
> @wujingyue, could you take the maintainer review here, including the end-to-end clip-norm behavior for the local and Transformer Engine attention paths? 

The standard process is to get this expert-reviewed by @NVIDIA/transformer and then final-reviewed. Will you or @NVIDIA/mcore-oncall figure out the right person?

**Comment by Connor-XY** [comment]
Thanks for the clarification, @wujingyue. I've requested @guihong-nv from the Transformer team for the expert review.

@guihong-nv, could you review the head-sharded softmax-offset metadata and end-to-end TP clipping norm for both local and Transformer Engine attention? The existing numerical report covers CPU/Gloo; the TE path still needs validation. This is finding 8 of #7452. We'll seek final review after the expert review.
